### PR TITLE
[thrift]try to fix bug.large amount of list or map values ,when socket can not read all and frame.remain still be decreased

### PR DESCRIPTION
**Pull request by pyrl247**
Description:
When client build large amount list or map values ,and socket can not read them all at once, and the state machine just decode half of list values ,if frame.remaining be decreased ,it will make list could not get all values(just got length - 1),and it will make state machine fall into disorder.

I use index rather then reference because the stack_ vector may expand and make the reference(got from stack_.back()) become invalid.
Risk Level:Medium
Testing:N/A
Docs Changes:N/A
Release Notes:N/A
Signed-off-by:Guang Yang pyrl247@gmail.com

### CODE REVIEW

**Comment by zuercher** [comment]
Thanks! There are a couple of changes we'll need before we can continue.

1.You'll need to modify your PR to to indicate you certify the DCO (see https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#dco-sign-your-work and https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco). In this case, if. you can certify compliance with the DCO, it should be sufficient to run `git commit --signoff --amend --no-edit && git push -f` to fix the commit. Just make sure to `--signoff` and further commits.

2. Please add a test that triggers the condition you're fixing so that we can avoid any future regressions. If you have questions about how to do that, post them here and I'm happy to help identify a test you can use as a blueprint.

**Comment by pyrl247** [comment]
I may have some problem about how to add test ,this is my first pr to this project.

**Comment by zuercher** [comment]
In `test/extensions/filters/network/thrift_proxy/decoder_test.cc`, you want to look at the test named `NoFieldValueData`. That's testing what happens when the state machine goes to read a struct field and there's no data available: passing result = false to the expectValue function simulates the case where there's insufficient data in the buffer to read the value.

The test `NoListValueData` tests the beginning of this for list values but doesn't test what happens when the value data arrives (which is where you fixed the bug). I think you can add expectations that it tries to read the list value again on a subsequent call to `dsm.run` which should exercise your code.

These tests mock out the protocol (which is what indicates that a buffer has insufficient data to continue), so we don't need to actually create buffers with many list values.

There are similar tests for maps ('NoMapKeyData' and `NoMapValueData`) and sets (`NoSetValueData`) that would cover your changes.

**Comment by zuercher** [comment]
And in fact, that points out to me that you need to implement your fix for map keys as well.

File: test/extensions/filters/network/thrift_proxy/decoder_test.cc [hunk 1]
```diff
@@ -336,6 +336,28 @@ TEST_P(DecoderStateMachineValueTest, ListValue) {
...
+
+  expectValue(proto_, handler_, field_type, false);
+
+  DecoderStateMachine dsm(proto_, metadata_, handler_);
+
+  dsm.setCurrentState(ProtocolState::ListBegin);
+  EXPECT_EQ(dsm.run(buffer), ProtocolState::WaitForData);
+  EXPECT_EQ(dsm.currentState(), ProtocolState::ListValue);
+
+  expectValue(proto_, handler_, field_type, false);
```
Comment (zuercher) [inline test/extensions/filters/network/thrift_proxy/decoder_test.cc:355]: I think you add an iteration where the value is present (result = true) to prove that the decoder continues and completes reading the list. (And similar steps in the other tests.)

File: test/extensions/filters/network/thrift_proxy/decoder_test.cc [hunk 2]
```diff
@@ -448,6 +470,31 @@ TEST_P(DecoderStateMachineValueTest, MapValueValue) {
   EXPECT_EQ(dsm.currentState(), ProtocolState::MapEnd);
 }
 
+TEST_P(DecoderStateMachineValueTest, IncompleteMapValues) {
```
Comment (zuercher) [inline test/extensions/filters/network/thrift_proxy/decoder_test.cc:473]: We need an Incomplete MapKeys test that performs this test again the map key (and matching code in the decoder to handle this case).

**Comment by rgs1** [comment]
> cc @rgs1 do you mind having a quick look?

Looking.

File: source/extensions/filters/network/thrift_proxy/decoder.cc [hunk 3]
```diff
@@ -111,13 +111,16 @@ DecoderStateMachine::DecoderStatus DecoderStateMachine::listBegin(Buffer::Instan
 // ListValue -> ListEnd
 DecoderStateMachine::DecoderStatus DecoderStateMachine::listValue(Buffer::Instance& buffer) {
   ASSERT(!stack_.empty());
-  Frame& frame = stack_.back();
-  if (frame.remaining_ == 0) {
+  uint32_t index = stack_.size() - 1;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/decoder.cc:114]: nit: `const uint32_t index`

File: source/extensions/filters/network/thrift_proxy/decoder.cc [hunk 4]
```diff
@@ -159,11 +162,14 @@ DecoderStateMachine::DecoderStatus DecoderStateMachine::mapKey(Buffer::Instance&
 // MapValue -> MapKey
 DecoderStateMachine::DecoderStatus DecoderStateMachine::mapValue(Buffer::Instance& buffer) {
   ASSERT(!stack_.empty());
-  Frame& frame = stack_.back();
-  ASSERT(frame.remaining_ != 0);
-  frame.remaining_--;
+  uint32_t index = stack_.size() - 1;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/decoder.cc:165]: ditto

File: source/extensions/filters/network/thrift_proxy/decoder.cc [hunk 5]
```diff
@@ -193,13 +199,16 @@ DecoderStateMachine::DecoderStatus DecoderStateMachine::setBegin(Buffer::Instanc
 // SetValue -> SetEnd
 DecoderStateMachine::DecoderStatus DecoderStateMachine::setValue(Buffer::Instance& buffer) {
   ASSERT(!stack_.empty());
-  Frame& frame = stack_.back();
-  if (frame.remaining_ == 0) {
+  uint32_t index = stack_.size() - 1;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/decoder.cc:202]: ditto

**Pull request merged**